File: forensics-btmgen/src/main/java/de/burger/forensics/plugin/GlobUtils.java
// DEST: src/main/java/de/burger/forensics/plugin/GlobUtils.java
package de.burger.forensics.plugin;

import java.nio.file.FileSystems;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.nio.file.Paths;
import java.util.concurrent.ConcurrentHashMap;
import java.util.regex.Pattern;
import org.jetbrains.annotations.NotNull;

public final class GlobUtils {

    private static final ConcurrentHashMap<String, Pattern> GLOB_CACHE = new ConcurrentHashMap<>();
    private static final ConcurrentHashMap<String, PathMatcher> PATH_MATCHER_CACHE = new ConcurrentHashMap<>();

    private GlobUtils() {
        throw new AssertionError("No instances.");
    }

    @NotNull
    public static String globToRegex(@NotNull String glob) {
        StringBuilder sb = new StringBuilder("^");
        int i = 0;
        while (i < glob.length()) {
            char c = glob.charAt(i);
            switch (c) {
                case '*':
                    if (i + 1 < glob.length() && glob.charAt(i + 1) == '*') {
                        sb.append(".*");
                        i++;
                    } else {
                        sb.append("[^/]*");
                    }
                    break;
                case '?':
                    sb.append('.');
                    break;
                case '.':
                case '(':
                case ')':
                case '+':
                case '|':
                case '^':
                case '$':
                case '@':
                case '%':
                    sb.append('\\').append(c);
                    break;
                case '{':
                    sb.append('(');
                    break;
                case '}':
                    sb.append(')');
                    break;
                case ',':
                    sb.append('|');
                    break;
                case '[':
                    sb.append('[');
                    break;
                case ']':
                    sb.append(']');
                    break;
                default:
                    sb.append(c);
                    break;
            }
            i++;
        }
        sb.append('$');
        return sb.toString();
    }

    @NotNull
    public static Pattern globToRegexCached(@NotNull String glob) {
        return GLOB_CACHE.computeIfAbsent(glob, key -> Pattern.compile(globToRegex(key)));
    }

    public static boolean globMatchesPath(@NotNull String relPathUnix, @NotNull String glob) {
        String separator = FileSystems.getDefault().getSeparator();
        String relPathPlatform = separator.equals("/") ? relPathUnix : relPathUnix.replace('/', separator.charAt(0));
        PathMatcher matcher = PATH_MATCHER_CACHE.computeIfAbsent(glob, key -> {
            String pattern = separator.equals("/") ? key : key.replace('/', separator.charAt(0));
            return FileSystems.getDefault().getPathMatcher("glob:" + pattern);
        });
        Path path = Paths.get(relPathPlatform);
        return matcher.matches(path);
    }
}

File: forensics-btmgen/src/main/kotlin/de/burger/forensics/plugin/GlobUtils.kt
package de.burger.forensics.plugin

import java.nio.file.FileSystems
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.PathMatcher
import java.util.concurrent.ConcurrentHashMap

private val globCache = ConcurrentHashMap<String, Regex>()
private val pathMatcherCache = ConcurrentHashMap<String, PathMatcher>()

// This must be pure; never call globToRegexCached() from here. No caching, no indirect recursion.
fun globToRegex(glob: String): String {
    val sb = StringBuilder("^")
    var i = 0
    while (i < glob.length) {
        when (val c = glob[i]) {
            '*' -> {
                if (i + 1 < glob.length && glob[i + 1] == '*') {
                    sb.append(".*")
                    i++
                } else {
                    sb.append("[^/]*")
                }
            }
            '?' -> sb.append('.')
            '.', '(', ')', '+', '|', '^', '$', '@', '%' -> sb.append('\\').append(c)
            '{' -> sb.append('(')
            '}' -> sb.append(')')
            ',' -> sb.append('|')
            '[' -> sb.append('[')
            ']' -> sb.append(']')
            else -> sb.append(c)
        }
        i++
    }
    sb.append('$')
    return sb.toString()
}

fun globToRegexCached(glob: String): Regex = globCache.computeIfAbsent(glob) { Regex(globToRegex(it)) }

fun globMatchesPath(relPathUnix: String, glob: String): Boolean {
    // Normalize '/' used in our code to the platform separator for PathMatcher
    val sep = FileSystems.getDefault().separator
    val relPathPlatform = if (sep == "/") relPathUnix else relPathUnix.replace('/', sep[0])
    val matcher = pathMatcherCache.computeIfAbsent(glob) {
        val pattern = if (sep == "/") glob else glob.replace('/', sep[0])
        FileSystems.getDefault().getPathMatcher("glob:" + pattern)
    }
    val p: Path = Paths.get(relPathPlatform)
    return matcher.matches(p)
}
